### PR TITLE
csi: add support for groupSnapshots in csi

### PR DESCRIPTION
**Pull request by raunakkumar**
- groupSnapshotter creates snapshots across all the PVCs in the group.
- since the snapshot is already created by the storage provider, the
corresponding request from external-snapshotter needs to be intercepted
and returned instead of creating it again

Signed-off-by: Raunak <raunak.kumar@hpe.com>

### CODE REVIEW

File: pkg/flavor/kubernetes/flavor.go [hunk 1]
```diff
@@ -339,6 +398,49 @@ func (flavor *Flavor) watchAllClaims(options meta_v1.ListOptions) (watch.Interfa
...
+	if flavor.snapshotIndexer == nil {
+		return nil, fmt.Errorf("requested snapshot %s was not found because snapshotIndexer was nil", name)
+	}
+
+	uid := name[9:len(name)] // snapshot-<uid>
+	log.Infof("Looking up VolumeSnapshot with uid %s", uid)
+
+	snapshots, err := flavor.snapshotIndexer.ByIndex("uid", uid)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve snapshot by uid %s", uid)
```
Comment (rkumpf) [inline pkg/flavor/kubernetes/flavor.go:414]: Include the error message we received?

File: pkg/flavor/kubernetes/flavor.go [hunk 2]
```diff
@@ -339,6 +398,49 @@ func (flavor *Flavor) watchAllClaims(options meta_v1.ListOptions) (watch.Interfa
...
+	} else if len(snapshots) == 0 {
+		return nil, fmt.Errorf("Requested snapshot %s was not found with uid %s", name, uid)
+	}
+
+	log.Infof("Found the following snapshots: %#v", snapshots)
```
Comment (rkumpf) [inline pkg/flavor/kubernetes/flavor.go:419]: Maybe "Found %#v snapshot(s)"?  (It should always be 1, right?)

File: pkg/flavor/types.go [hunk 3]
```diff
@@ -35,4 +35,5 @@ type Flavor interface {
 	GetNFSVolumeID(volumeID string) (string, error)
 	GetOrchestratorVersion() (*version.Info, error)
 	MonitorPod(podLabelkey, podLabelvalue string) error
+	GetSnapshotNameFromSnapshot(snapshotName string) (string, error)
```
Comment (rkumpf) [inline pkg/flavor/types.go:38]: This function name is kind of hard to understand.  It seems like it should take a snapshot parameter.  Should it be something like GetGroupSnapshotNameFromSnapshotName?

File: pkg/flavor/kubernetes/flavor.go [hunk 4]
```diff
@@ -339,6 +398,49 @@ func (flavor *Flavor) watchAllClaims(options meta_v1.ListOptions) (watch.Interfa
 	return flavor.kubeClient.CoreV1().PersistentVolumeClaims(meta_v1.NamespaceAll).Watch(options)
 }
 
+func (flavor *Flavor) getSnapshotFromSnapshotName(name string) (*v1beta1.VolumeSnapshot, error) {
+	log.Infof(">>>>> getSnapshotFromSnapshotName called with %s", name)
```
Comment (rkumpf) [inline pkg/flavor/kubernetes/flavor.go:402]: Should be Tracef?

File: pkg/flavor/kubernetes/flavor.go [hunk 5]
```diff
@@ -339,6 +398,49 @@ func (flavor *Flavor) watchAllClaims(options meta_v1.ListOptions) (watch.Interfa
...
+		return nil, fmt.Errorf("Requested snapshot %s was not found with uid %s", name, uid)
+	}
+
+	log.Infof("Found the following snapshots: %#v", snapshots)
+
+	return snapshots[0].(*v1beta1.VolumeSnapshot), nil
+}
+
+func (flavor *Flavor) GetSnapshotNameFromSnapshot(name string) (string, error) {
+	log.Infof(">>>>> GetSnapshotNameFromSnapshot called with snapshot %s", name)
```
Comment (rkumpf) [inline pkg/flavor/kubernetes/flavor.go:425]: Tracef

**Review by rkumpf — COMMENTED** [review]
Overall it looks good as far as I can tell.  A couple little things seem worth changing.

**Comment by raunakkumar** [comment]
Sure, will address these changes

**Review by rkumpf — APPROVED** [review]

File: pkg/driver/controller_server.go [hunk 6]
```diff
@@ -1240,6 +1240,13 @@ func (driver *Driver) CreateSnapshot(ctx context.Context, request *csi.CreateSna
 		delete(createOptions, descriptionKey)
 	}
 
+	// check if this snapshot is from group snapshotter
+	groupSnapshotName, err := driver.flavor.GetGroupSnapshotNameFromSnapshotName(request.Name)
```
[the remark this answers is not in the document]
Comment (raunakkumar) [reply pkg/driver/controller_server.go:1244]: If there is any error, we fallback to the existing snapshot workflow and not groupSnapshot

[the remark this answers is not in the document]
Comment (suneeth51) [reply pkg/driver/controller_server.go:1244]: Agreed, atleast log the error and add a comment.

**Review by suneeth51 — APPROVED** [review]

**Pull request merged**